internal/eventstream/client.go

```diff
@@ -0,0 +1,455 @@
+// derived from the Client.Ctx and would either stop when this context is done or when the context.CancelFunc is called.
+func (client *Client) startCatchupWorkers() (chan *eventMsg, context.CancelFunc) {
+	startTime := time.Now()
+	eventMsgCh := make(chan *eventMsg)
+
+	// Unfortunately, the errgroup context is hidden, that's why another context is necessary.
+	ctx, cancel := context.WithCancel(client.Ctx)
+	group, groupCtx := errgroup.WithContext(ctx)
+
+	objTypes := []string{"host", "service"}
+	for _, objType := range objTypes {
+		objType := objType // https://go.dev/doc/faq#closures_and_goroutines
+		group.Go(func() error {
+			err := client.checkMissedChanges(groupCtx, objType, eventMsgCh)
+			if err != nil {
+				client.Logger.Errorw("Catch-up-phase event worker failed", zap.String("object type", objType), zap.Error(err))
+			}
+			return err
+		})
+	}
+
+	go func() {
+		err := group.Wait()
+		if err != nil {
+			client.Logger.Errorw("Catching up the API failed", zap.Error(err), zap.Duration("duration", time.Since(startTime)))
+		} else {
+			client.Logger.Infow("Catching up the API has finished", zap.Duration("duration", time.Since(startTime)))
+		}
+
+		cancel()
+		close(eventMsgCh)
+	}()
+
+	return eventMsgCh, cancel
+}
+
+// worker is the Client's main background worker, taking care of event.Event dispatching and mode switching.
+//
+// When the Client is in the catch-up-phase, requested by catchupPhaseRequest, events from the Event Stream API will
+// be cached until the catch-up-phase has finished, while replayed events will be delivered directly.
+//
+// Communication takes place over the eventDispatcherEventStream and catchupPhaseRequest channels.
+func (client *Client) worker() {
+	// eventCache is a subset of event.Event used for caching those in catchupCache below.
+	type eventCache struct {
+		SourceId int64
+		Name     string
+		Type     string
+	}
+
+	var (
+		// catchupEventCh emits events generated during the catch-up-phase from catch-up-workers. It will be closed when
+		// catching up is done, which indicates the select below to switch phases. When this variable is nil, this
+		// Client is in the normal operating phase.
+		catchupEventCh chan *eventMsg
+		// catchupCancel cancels, if not nil, all running catch-up-workers, e.g., when restarting catching-up.
+		catchupCancel context.CancelFunc
+
+		// catchupBuffer holds Event Stream events to be replayed after the catch-up-phase has finished.
+		catchupBuffer = make([]*event.Event, 0)
+		// catchupCache maps eventCache derived from event.Events to API time to skip replaying outdated events.
+		catchupCache = make(map[eventCache]time.Time)
+
+		// dispatchQueue is a FIFO-like queue for events to be dispatched to the callback function without having to
+		// wait for the callback to finish, which, as being database-bound, might take some time during bulk phases.
+		dispatchQueue = make(chan *event.Event, 1<<16)
```

Comment:
I don't really get the idea behind this extra buffered channel. There's already `catchupBuffer` for buffering and combined with how it's used, it looks like it introduces some 65536 events + how many events can be handled within a minute limit.

Can you please explain what this is intended to do? I'm not really sure if it's supposed to smooth out some database latency, or take up the entire contents of `catchupBuffer` at times, or both.

Reply:
All events - fetched over the Event Stream API during normal operation, fetched from the Icinga 2 API during the catch-up-phase, and replayed Event Stream API events when leaving the catch-up-phase - are being sent to the `dispatchQueue` for processing. This is done through the `dispatchEvent` helper, which also checks the `context.Context` and has an additional huge timeout if there is very much to do or something went really wrong.

<details>
<summary>Code snippets writing into dispatchQueue</summary>

https://github.com/Icinga/icinga-notifications/blob/9ccd2c28d7b7240dd47e26118d93ba532dc4aa93/internal/eventstream/client.go#L344-L359
https://github.com/Icinga/icinga-notifications/blob/9ccd2c28d7b7240dd47e26118d93ba532dc4aa93/internal/eventstream/client.go#L391-L397
https://github.com/Icinga/icinga-notifications/blob/9ccd2c28d7b7240dd47e26118d93ba532dc4aa93/internal/eventstream/client.go#L401-L414
https://github.com/Icinga/icinga-notifications/blob/9ccd2c28d7b7240dd47e26118d93ba532dc4aa93/internal/eventstream/client.go#L422-L430
</details>

In parallel, another goroutine consumes events from the buffered `dispatchQueue` goroutine and sends them sequentially to the defined callback function.

<details>
<summary>Consuming goroutine around dispatchQueue</summary>

https://github.com/Icinga/icinga-notifications/blob/9ccd2c28d7b7240dd47e26118d93ba532dc4aa93/internal/eventstream/client.go#L330-L342
</details>

The `catchupBuffer`, on the other hand, is only used to delay Event Stream events during the catch-up-phase and gets filled until the phase ends and is only then being consumed.

As the `dispatchQueue` covers all events (see above) and is buffered with an huge buffer of up to 2^16 possible events, all three possible producers should not block, or under an immense load only for a very short time, when enqueuing an event to be delivered to the callback.

In contrast, the prior and unbuffered implementation which just sent events directly to the callback function always blocked for a short amount of time. For just a few elements, this delay was too short to be relevant. However, in my test instance with round about 4k initial state changes being sent to a Rocket.Chat, this delay summed up to over a minute due to database queries and interacting with the chat.

By the way, I made those change in regard to your earlier comment, <https://github.com/Icinga/icinga-notifications/pull/112#discussion_r1385106470>.

I hope it could made my motivations clear. Please feel free to argue on them if I'm mistaking something or if something is still unclear.

Reply:
> By the way, I made those change in regard to your earlier comment, [#112 (comment)](https://github.com/Icinga/icinga-notifications/pull/112#discussion_r1385106470).

The first idea I had in mind was to rewrite the following case (code from the PR when that comment was written)

https://github.com/Icinga/icinga-notifications/blob/500b2a974d5c03d6e811735d62d3db0f6a433d93/internal/eventstream/client.go#L348-L366

something like this (reduced to the parts relevant for what I'm trying to say, so missing check of `replayCache`, logging, reset of other variable):

```go
case <-replayFinCh:
    // process a single event
    client.CallbackFn(replayBuffer[0])
    replayBuffer = replayBuffer[1:]

    if len(replayBuffer) == 0 {
        // no more events left, disable this case
        replayFinCh = nil
    } else {
        // nothing explicit to do here, implicitly continue the loop,
        // allowing the other cases to add another event to replayBuffer
    }
```

That on its own probably isn't perfect. For one, it heavily relies on the behavior when multiple cases are ready and still blocks the loop, but limited to the processing time of a single event now. While there's room for improvement, it should convey the idea.

With the current PR, there are these extra 65536 and 1 minute limits, the need to clear an extra buffer, and there are cases where it's ignored that events are dismissed. For now, this looks like complexity where I don't see what it's for, as there should be no need for a second stage of buffering and this all seems to exist just for managing that second buffer.

Reply:
As I felt unsure, I just "verified" the fairness of the current implementation within the debugger. It behaves as the [Go Language Specification says about `select`](https://go.dev/ref/spec#Select_statements) that "[i]f one or more of the communications can proceed, a single one that can proceed is chosen via a uniform pseudo-random selection". Just to be sure, I also wrote a [small mockup](https://go.dev/play/p/qhNf6C_LM1p) which shows - please note that the runs are not deterministic, as stated in the specs above - that the other producer may also be selected when working on the closed producer.